mantle/kola/harness.go

```diff
@@ -564,7 +564,7 @@ func runProvidedTests(tests map[string]*register.Test, patterns []string, multip
 
 	var nonExclusiveTests []*register.Test
 	for _, test := range tests {
-		if test.NonExclusive {
+		if test.NonExclusive && !disableNonExclusive {
```

Comment:
Hmm, not sure I follow. Why do we not want the non-exclusive logic when rerunning?

Reply:
I think this was originally implemented because of a suggestion made by me. The idea is that if a quote/unquote "non-exclusive test" failed then we wanted to verify the test also fails when run exclusively. This is an indicator of if the test actually has some conflict and needs to be marked as exclusive (or the other conflicting test needs to be marked as exclusive).

Reply:
Ahhh OK. Hmm, overall I think it'd be less surprising if `kola rerun` reran tests in the same way they were originally run. Maybe instead we can add a `--disable-non-exclusive` switch which is generally useful and then allows a developer to enable that manually when running either `kola run` or `kola rerun`. WDYT?

Reply:
> Maybe instead we can add a --disable-non-exclusive switch which is generally useful and then allows a developer to enable that manually when running either kola run or kola rerun.

I think this is a good option especially since this functionality isn't documented. I think one would need to be following the non-exclusive and timeout work closely to understand why we want to rerun tests in separate VMs. So turning of that behaviour by default would probably make the reruns easier to follow.

Reply:
> Ahhh OK. Hmm, overall I think it'd be less surprising if `kola rerun` reran tests in the same way they were originally run. Maybe instead we can add a `--disable-non-exclusive` switch which is generally useful and then allows a developer to enable that manually when running either `kola run` or `kola rerun`. WDYT?

yeah I'm not entirely sure. the problem is that kola is going to rerun only the tests that fail, so it's already not re-running the tests in exactly the same way it ran them the first time. I kind of like knowing each test gets its own isolated env in the re-run.

If we decide to change this behavior I don't really think we need a `--disable-non-exclusive` flag. The developer can just do that manually by specifying a single test to run.

Reply:
> yeah I'm not entirely sure. the problem is that kola is going to rerun only the tests that fail, so it's already not re-running the tests in exactly the same way it ran them the first time. I kind of like knowing each test gets its own isolated env in the re-run.

Don't feel strongly about it. But again, I think this violates the principle of least surprise. E.g. the test names in the output will be different. As Saqib correctly points out, it'll only be understood to a few folks why things are subtly different. Rerunning tests IMO should be semantically equivalent to `kola run $list_of_tests_which_fail`.

> If we decide to change this behavior I don't really think we need a `--disable-non-exclusive` flag. The developer can just do that manually by specifying a single test to run.

Not today. I agree it probably should but also OK either way. See discussions in https://github.com/coreos/coreos-assembler/pull/2531.